qiskit/transpiler/passes/mapping/barrier_before_final_measurements.py

```diff
@@ -92,17 +93,19 @@ def run(self, dag):
                     and our_ancestors.isdisjoint(their_descendants)
                     and our_descendants.isdisjoint(their_ancestors)
             ):
-                merge_barrier = Barrier(qubits=(our_qubits | their_qubits))
-                merge_barrier_node = barrier_layer.apply_operation_front(merge_barrier)
+                barrier_to_add = Barrier(qubits=(our_qubits | their_qubits))
 
                 our_ancestors = our_ancestors | their_ancestors
                 our_descendants = our_descendants | their_descendants
 
                 barrier_layer._remove_op_node(candidate_barrier)
-                barrier_layer._remove_op_node(new_barrier_node)
 
-                new_barrier_node = merge_barrier_node
+                if new_barrier_node:
+                    barrier_layer._remove_op_node(new_barrier_node)
+                    new_barrier_node = None
```

Comment:
It might be clearer to push this outside the `for candidate_barrier in existing_barriers:` loop. That may give a cleaner logical separation between the 'analyze the dag for redundant barriers' stage and the 'replace the newly added barrier with its equivalent barrier' stage.